source/Detector/Amptek/AmptekSDD123Server.cpp

```diff
@@ -0,0 +1,570 @@
+	}
+}
+
+bool AmptekSDD123Server::detectorUnavailable() const{
+	return detectorUnavailable_;
+}
+
+QList<double> AmptekSDD123Server::droppedPacketStatistics() const{
+	QDateTime last30Seconds = QDateTime::currentDateTime().addSecs(-30);
+	QList<double> retVal;
+	retVal << droppedPackets_.totalOccurances() << droppedPackets_.totalRate() << droppedPackets_.occurancesSince(last30Seconds) << droppedPackets_.rateSince(last30Seconds);
+	return retVal;
+}
+
+QList<double> AmptekSDD123Server::unrequestedPacketStatistics() const{
+	QDateTime last30Seconds = QDateTime::currentDateTime().addSecs(-30);
+	QList<double> retVal;
+	retVal << unrequestedPackets_.totalOccurances() << unrequestedPackets_.totalRate()<< unrequestedPackets_.occurancesSince(last30Seconds) << unrequestedPackets_.rateSince(last30Seconds);
+	return retVal;
+}
+
+bool AmptekSDD123Server::event(QEvent *e){
+	if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationInitiateRequestEvent){
+		if(((AmptekConfigurationInitiateRequestEvent*)e)->configurationMode_){
+			emit serverAboutToChangeToConfigurationState();
+			QTimer::singleShot(requestIntervalTimer_+5, this, SLOT(disableMCAMCS()));
+		}
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationRequestEvent){
+		if(((AmptekConfigurationRequestEvent*)e)->allParametersRequest_)
+			requestAllTextConfigurationParameters();
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::DwellRequestEvent){
+		if(((AmptekDwellRequestEvent*)e)->dwellMode_)
+//			enableMCAMCS();
+			requestDataPacket(AmptekCommandManagerSGM::RequestEnableMCAMCS);
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationSetEvent){
+		QStringList configurationCommands = ((AmptekConfigurationSetEvent*)e)->configurationCommands_;
+		textConfigurationSet(configurationCommands.at(0));
+		e->accept();
+		return true;
+	}
+	return QObject::event(e);
+}
+
+int AmptekSDD123Server::forwardDatagram(const QByteArray &datagram){
+	AmptekSDD123Packet forwardingPacket(packetIDCounter_++, datagram, true);
+	if(forwardingPacket.isValid()){
+		sendRequestDatagram(forwardingPacket);
+		return forwardingPacket.packetID();
+	}
+	return -1;
+}
+
+void AmptekSDD123Server::requestDataPacket(AmptekCommandManagerSGM::AmptekCommandDefSGM command, const QString &dataStringHex, bool sendFakeRequest, bool sendSyncRequest, int overrideTimeout)
+{
+	AmptekSDD123Packet requestStatusPacket(packetIDCounter_++, AmptekCommandManagerSGM::sddCommands()->amptekCommand(command).hex(), dataStringHex);
+	if (sendFakeRequest) {
+		fakeSendDatagram(requestStatusPacket, overrideTimeout);
+	} else if (sendSyncRequest) {
+		requestStatusPacket.appendRelatedSyncRequestID(timeOutIDCounter_);
+		sendSyncDatagram(requestStatusPacket, overrideTimeout);
+	} else {
+		sendRequestDatagram(requestStatusPacket);
+	}
+}
+
+//void AmptekSDD123Server::requestStatus(){
```

Comment:
These commented codes/functions will be deleted before merge.

I created a function ```requestDataPacket()``` to replace all those functions. I keep those replaced codes here for the convenience of reviewers to understand the functionality of ```requestDataPacket()```.